File: src/main/java/ua/nure/fedorenko/kidstim/model/entity/Reward.java
package ua.nure.fedorenko.kidstim.model.entity;

import com.fasterxml.jackson.annotation.JsonIgnore;
import org.hibernate.annotations.GenericGenerator;

import javax.persistence.*;
import java.io.Serializable;
import java.util.List;

@Entity
@Table
public class Reward implements Serializable {

    @Id
    @GeneratedValue(generator = "uuid")
    @GenericGenerator(name = "uuid", strategy = "uuid")
    @Column(name = "id")
    private String id;

    @Column(name = "description")
    private String description;

    @Column(name = "points")
    private int points;

    @Enumerated(EnumType.ORDINAL)
    @Column(name = "status")
    private RewardStatus status;

    @ManyToMany(fetch = FetchType.LAZY, cascade = {CascadeType.MERGE, CascadeType.REFRESH})
    @JoinTable(name = "child_reward", joinColumns = {@JoinColumn(name = "reward_id")}, inverseJoinColumns = {@JoinColumn(name = "child_id")})
    private List<Child> children;

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "parent")
    private Parent parent;


    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public RewardStatus getStatus() {
        return status;
    }

    public void setStatus(RewardStatus status) {
        this.status = status;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public int getPoints() {
        return points;
    }

    public void setPoints(int points) {
        this.points = points;

    }

    public List<Child> getChildren() {
        return children;
    }

    public void setChildren(List<Child> children) {
        this.children = children;
    }

    public Parent getParent() {
        return parent;
    }

    public void setParent(Parent parent) {
        this.parent = parent;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Reward)) return false;

        Reward reward = (Reward) o;

        if (getPoints() != reward.getPoints()) return false;
        if (getDescription() != null ? !getDescription().equals(reward.getDescription()) : reward.getDescription() != null)
            return false;
        if (getStatus() != reward.getStatus()) return false;
        return getParent().equals(reward.getParent());
    }

    @Override
    public int hashCode() {
        int result = getDescription() != null ? getDescription().hashCode() : 0;
        result = 31 * result + getPoints();
        result = 31 * result + (getStatus() != null ? getStatus().hashCode() : 0);
        result = 31 * result + getParent().hashCode();
        return result;
    }
}

File: src/main/java/ua/nure/fedorenko/kidstim/model/entity/Reward.kt
package ua.nure.fedorenko.kidstim.model.entity

import java.io.Serializable
import org.hibernate.annotations.GenericGenerator
import javax.persistence.*

@Entity
@Table
data class Reward(
        @Id
        @GeneratedValue(generator = "uuid")
        @GenericGenerator(name = "uuid", strategy = "uuid")
        @Column(name = "id")
        var id: String? = null,

        @Column(name = "description")
        var description: String? = null,

        @Column(name = "points")
        var points: Int = 0,

        @Enumerated(EnumType.ORDINAL)
        @Column(name = "status")
        var status: RewardStatus? = null
) : Serializable {


    @ManyToMany(fetch = FetchType.LAZY, cascade = [(CascadeType.MERGE), (CascadeType.REFRESH)])
    @JoinTable(name = "child_reward", joinColumns = [(JoinColumn(name = "reward_id"))], inverseJoinColumns = [(JoinColumn(name = "child_id"))])
    var children: List<Child>? = ArrayList()

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "parent")
    var parent: Parent? = null
}